firestore/bulkwriter.go

```diff
@@ -0,0 +1,325 @@
+		}
+		return writeResultFromProto(bwr.result)
+	}
+}
+
+// setError ensures that an error is returned on the error channel of BulkWriterJob.
+func (j *BulkWriterJob) setError(e error) {
+	bwr := bulkWriterResult{
+		err:    e,
+		result: nil,
+	}
+	j.resultChan <- bwr
+	close(j.resultChan)
+}
+
+// A BulkWriter supports concurrent writes to multiple documents. The BulkWriter
+// submits document writes in maximum batches of 20 writes per request. Each
+// request can contain many different document writes: create, delete, update,
+// and set are all supported.
+//
+// Only one operation (create, set, update, delete) per document is allowed.
+// BulkWriter cannot promise atomicity: individual writes can fail or succeed
+// independent of each other. Bulkwriter does not apply writes in any set order;
+// thus a document can't have set on it immediately after creation.
+type BulkWriter struct {
+	database        string           // the database as resource name: projects/[PROJECT]/databases/[DATABASE]
+	start           time.Time        // when this BulkWriter was started; used to calculate qps and rate increases
+	vc              *vkit.Client     // internal client
+	maxOpsPerSecond int              // number of requests that can be sent per second
+	docUpdatePaths  map[string]bool  // document paths with corresponding writes in the queue
+	limiter         rate.Limiter     // limit requests to server to <= 500 qps
+	bundler         *bundler.Bundler // handle bundling up writes to Firestore
+	ctx             context.Context  // context for canceling all BulkWriter operations
+	openLock        sync.Mutex       // guards against setting isOpen concurrently
+	isOpen          bool             // flag that the BulkWriter is closed
+}
+
+// newBulkWriter creates a new instance of the BulkWriter.
+func newBulkWriter(ctx context.Context, c *Client, database string) *BulkWriter {
+	// Although typically we shouldn't store Context objects, in this case we
+	// need to pass this Context through to the Bundler handler.
+	ctx = withResourceHeader(ctx, c.path())
+
+	bw := &BulkWriter{
+		database:        database,
+		start:           time.Now(),
+		vc:              c.c,
+		isOpen:          true,
+		maxOpsPerSecond: defaultStartingMaximumOpsPerSecond,
+		docUpdatePaths:  make(map[string]bool),
+		ctx:             ctx,
+		limiter:         *rate.NewLimiter(rate.Limit(maxWritesPerSecond), 1),
+	}
+
+	// can't initialize within struct above; need instance reference to BulkWriter.send()
+	bw.bundler = bundler.NewBundler(&BulkWriterJob{}, bw.send)
+	bw.bundler.HandlerLimit = bw.maxOpsPerSecond
+	bw.bundler.BundleCountThreshold = maxBatchSize
+
+	return bw
+}
+
+// End sends all enqueued writes in parallel and closes the BulkWriter to new requests.
+// After calling End(), calling any additional method automatically returns
+// with an error. This method completes when there are no more pending writes
+// in the queue.
+func (bw *BulkWriter) End() {
+	bw.Flush()
```

Comment:
Shouldn't this should go after setting `isOpen` to false (e.g., prevent additional write requests, and then flush)?

Reply:
Done.